### ISSUE TITLE
dart:io stdin is a non-broadcast stream

### ISSUE THREAD

**Issue by BlackHC** [dart-lang/sdk#24881]
How exactly is this supposed to work? A global stream that only supports a single listener is not very scale-able. I'm trying to run a process interactively from within dart and this makes it very hard to do that reliably.

What am I missing? It seems that the current dart:io implementation does not make sense.

Thanks,
 Andreas

**Comment by lrhn** [comment]
There is only one stdin in the process, and when you consume input, it's gone.

If you want multiple concurrent listeners on input, you can use `Stream.asBroadcastStream()`.

If you want to look at the input and do things depending on the input, you can look at `StreamQueue` in `package:async` which makes it easier to act on a stream one event at a time.

**Comment by BlackHC** [comment]
tl;dr: Having a global non-broadcast stream is still stupid and your design is inconsistent. Your answer is an obvious workaround but it does not fix the underlying conceptual issue at hand

Maybe you have misunderstood me. Having a global stdin variable means that is is impossible for a library to use it by itself because that code cannot know if stdin has already been listened to or not.

Moreever, readLineSync and readByteSync don't have this issue at all. You can call them from anywhere. This is inconsistent.

How is anyone supposed to use the "Stream" stdin at all in a library/shared code? I guess the answer is not at all? Is this part of Dart a priority at all or should we look at other scripting languages in general for console scripting?

**Comment by lrhn** [comment]
You can't really use stdin from a library, for the reasons you state.
I think that's OK. The library should not access stdin directly, it should take a stream as a parameter, and then the user can choose to use stdin - or some other stream. 

Making the stream broadcast isn't going to make anything *better*. Instead you'll likely find that some of the input has been consumed before you even get to listen to it. A library can still not assume that it can listen to stdin - it has no idea where it is in the stdin stream.

The stdin stream is a "generated" stream. It can't be replayed, it can't be rewound, when you use it it is consumed in the process. the stream's events form a consecutive whole, and dropping events leaves you in an uncertain state (maybe in the middle of a line).
Streams like that are always single-subscription streams.

**Comment by BlackHC** [comment]
Thanks for the explanation! I'll try to do that then and use StreamQueue.
